### ISSUE TITLE
Consider adding tag size of 4 and 8 to AES GCM

### ISSUE THREAD

**Issue by d3lm** [RustCrypto/AEADs#541]
I have noticed that `AesGcm` only accepts tag sizes [>= 12 and <= 16](https://github.com/RustCrypto/AEADs/blob/master/aes-gcm/src/lib.rs#L152-L156). However, Node.js for example also allows tag sizes of 4 and 8 (see [here](https://github.com/nodejs/node/blob/16b8c71a324be524d44d48d945a2159abb94528b/src/crypto/crypto_cipher.cc#L49-L51)). Could we implement `SealedTagSize` for `consts::U4` and `consts::U8` as well?

**Comment by d3lm** [comment]
Oh ok, I wasn't actually aware of that - thanks for pointing that out. I guess it'd be more effort than to support short tags. It might be ok for us to restrict the tag size and ignore short tags. I also don't know if there's applications that really use short tags and it's just something I noticed while I was implementing `AES-GCM` that complies to Node.js using `aes_gcm`.

**Comment by d3lm** [comment]
Is it possible for me to compute the authentication tag manually? I am asking because in Node.js you can encrypt multiple times by calling `update(<data>[, encoding])` and only after calling `final()` the authentication tag is computed over all encrypted data.

**Comment by d3lm** [comment]
What I don't fully understand with how it's done in Node.js is that normally the nonce / IV should not be re-used for multiple pieces of data. AFAIU if you were to use the same nonce for the different pieces of data you're encrypting with AES-GCM, you would severely compromise the security of your encrypted data. Therefore, the nonce should be unique for each piece of data being encrypted with the same key. Isn't that true? Wonder why Node.js then allows you to call `update` multiple times.

**Comment by tarcieri** [comment]
It sounds like they are buffering and perhaps incrementally encrypting/authenticating the data.

We only have one-shot APIs at the moment, though you could largely achieve the same effect by buffering the data into a `Vec<u8>` and then encrypting it with the one-shot API.

**Comment by d3lm** [comment]
Yes, that's what I am doing for encryption and it works, but I have issues with incrementally decrypt data. Because parts of the encrypted data will throw an `aead::Error` likely because the tag isn't correct and corresponds to the tag for the entire data. Any idea what I can do?

**Comment by tarcieri** [comment]
It's not possible to safely decrypt individual AEAD messages incrementally. At the very least, the tag needs to be checked before decryption can begin.

**Comment by d3lm** [comment]
I wonder how Node.js does that then 🤔

**Comment by tarcieri** [comment]
My guess would be: unsafely

**Comment by d3lm** [comment]
Oh my, I just realized, with every call to `update(<partial_aead_message>)` they don't even check the tag only once you call final and have all the "chunks".

**Comment by d3lm** [comment]
But that is prolly not a good idea is it?

**Comment by d3lm** [comment]
Hm I guess maybe it is because only after calling `final()` the entire decryption process is finished and I would expect that you can not expect it to be fully checked before calling `final()`.

**Comment by tarcieri** [comment]
Operating on unauthenticated data can be a source of chosen ciphertext attacks which completely undermine AEAD security.

These sorts of streaming decryption APIs are "hazmat" which is difficult to use correctly with high misuse potential.

**Comment by d3lm** [comment]
Thanks a lot for your input on this 🙏 I am not a crypto expert myself so appreciate it.

**Comment by d3lm** [comment]
Maybe you can still consider adding short tags, that would be really nice. But for now it's fine and I can work around it and only support 128 bit tags.

**Comment by tarcieri** [comment]
We could add an API for it, but it would require a special decryptor object which caps the max message size and number of decryptions allowed under a given key

**Comment by d3lm** [comment]
I think that sounds reasonable as long as those numbers are decently high or maybe configurable?

**Comment by d3lm** [comment]
I think generally I am fighting a bit the fact that the nonce length and tag length are both configured "statically" and I can't pass them as arguments dynamically.

**Comment by tarcieri** [comment]
Do you have a particular use case for dynamic tag lengths which isn't limited to a small number of possibilities?

**Comment by d3lm** [comment]
I think it's just the ability to pass in the tag length from the passed u8 slice for the tag instead of statically defining the tag length.

**Comment by d3lm** [comment]
The use case is that I can create the "handle" so `AesGcm::<Aes, NonceLength, TagSize>::new_from_slice(key)` once and not every time I encrypt partial data which is what I need to implement the same interface as Node.js. I suppose creating the handle is not expensive so it should be somewhat fine? It'd be really nice if it was possible to do
```
AesGcm::<Aes, NonceLength, TagSize>::new_from_slice(key, nonce_size, tag_size)
```
instead of passing the sizes as types. Curious to hear your thoughts because right now I have a `match` for the tag size but for nonce sizes it's a bit unfortunate because I'd have to define every size and then pass the right type.

**Comment by tarcieri** [comment]
Sorry, that's not what I'm asking. Why are you working with arbitrary tag sizes in the first place? Who is choosing them? Why are they arbitrary? Why aren't they some fixed, low-cardinality set as specified by a protocol you're implementing?

**Comment by d3lm** [comment]
The user is choosing the tag size. In Node.js you can do:

```js
const cipher = crypto.createCipheriv(algorithm, key, iv, { authTagLength: 4 });
```

So when the cipher object is created they can pass in the auth tag size. By default it's 16. Node allows 4, 8, and 12 - 16 byte auth tags.

**Comment by tarcieri** [comment]
I'm afraid that still doesn't answer my question.

I am asking for specific examples of scenarios where being able to dynamically select the IV size using a runtime parameter would be helpful.

You gave a code example which uses a constant literal. That's not only not an example of a use case/scenario, but also an example of the sort which is very much amenable to being a compile time constant.

**Comment by d3lm** [comment]
Hm I am not sure I understand. It's a constant literal in the user's code, but in our runtime we don't know those values upfront and we run arbitrary code that can use any IV length or different tag sizes.

**Comment by tarcieri** [comment]
So your use case is you're implementing a Node.js interpreter and need to support this kind of dynamic selection?

If so, please file a new issue specific to that request.

**Comment by d3lm** [comment]
Yea exactly. We are building a Node.js runtime for the browser so you can run Node.js entirely in your browser without relying on cloud servers. Should have said that in the beginning, sorry about that.

**Comment by mouse07410** [comment]
One example of where variable tag size may help is a noisy link, where you adjust the tag size based on the available practical bandwidth and the point in the protocol flow.

**Comment by kkoyung** [comment]
I am currently working on the WebCrypto API implementation in the Servo browser, which uses the `aes-gcm` crate to provide AES-GCM support. The WebCrypto specification (https://w3c.github.io/webcrypto/#aes-gcm-operations-encrypt) recommends that browsers support tag lengths of 32, 64, 96, 104, 112, 120, and 128 bits for compatibility reasons. Most major browsers follow the specification to support all of them. I believe this is why Node.js supports 32-bit and 64-bit authentication tags.

I would like to know whether RustCrypto is still considering adding support for 32-bit and 64-bit tag lengths.

**Comment by tarcieri** [comment]
@kkoyung does the WebCrypto specification actually follow [NIST SP 800-38D](https://csrc.nist.gov/publications/detail/sp/800-38d/final) in regards to these tags or is it in violation?

> For any implementation that supports 32-bit or 64-bit tags, one of the rows in Table 1 or Table 2,
respectively, shall be enforced. In particular, the supported lengths for the plaintext/ciphertext
and the AAD shall ensure that every valid packet satisfies the length restriction in the row, and
the controlling protocol/system shall ensure that the key is changed before the authenticated
decryption function is invoked more than the maximum that is given in the row. A smaller
maximum may also be enforced.

Supporting such short tags requires a stateful AEAD which keeps track of its number of invocations and errors out if it has been used too many times, which we don't currently implement or support.

I guess we could potentially add such lengths under a hazmat feature and document NIST SP 800-38D's usage requirements, but I'm worried if I do that people will just run off and use short tags without honoring the usage SP 800-38D requirements (which it seems may already be baked into the WebCrypto spec)

**Comment by newpavlov** [comment]
I think it's fine to add support for such tag sizes gated on a hazmat feature and document that it's user responsibility to satisfy the quoted requirement.
